src/main/java/betterquesting/handlers/EventHandler.java

```diff
@@ -417,10 +419,39 @@ public void onCommand(CommandEvent event) {
     private final ArrayDeque<EntityPlayerMP> opQueue = new ArrayDeque<>();
     private boolean openToLAN = false;
 
+    private static HashSet<EntityPlayer> playerInventoryUpdates = new HashSet<>();
+
+    /**
+     * Schedules checking player's inventory on the next server tick.
+     * Deduplicates requests to avoid scanning it multiple times per tick.
+     */
+    public static void schedulePlayerInventoryCheck(EntityPlayer player) {
+        synchronized (playerInventoryUpdates) {
+            playerInventoryUpdates.add(player);
+        }
+    }
+
     @SubscribeEvent
     public void onServerTick(ServerTickEvent event) {
-        if (event.phase == Phase.START && FMLCommonHandler.instance().getMinecraftServerInstance().getTickCounter() % 60 == 0) {
-            AdvListenerManager.INSTANCE.updateAll();
+        if (event.phase == Phase.START) {
+            if (FMLCommonHandler.instance().getMinecraftServerInstance().getTickCounter() % 60 == 0) {
+                AdvListenerManager.INSTANCE.updateAll();
+            }
+            synchronized (playerInventoryUpdates) {
+                for (EntityPlayer player : playerInventoryUpdates) {
+                    if (player == null || player.inventory == null) {
+                        continue;
+                    }
+                    ParticipantInfo pInfo = new ParticipantInfo(player);
+
+                    for (DBEntry<IQuest> entry : QuestingAPI.getAPI(ApiReference.QUEST_DB).bulkLookup(pInfo.getSharedQuests())) {
+                        for (DBEntry<ITask> task : entry.getValue().getTasks().getEntries()) {
+                            if (task.getValue() instanceof ITaskInventory) ((ITaskInventory)task.getValue()).onInventoryChange(entry, pInfo);
+                        }
+                    }
+                }
+                playerInventoryUpdates.clear();
```

Comment:
I understand the `synchronized` block when scheduling, but when looping through here, it should be fine to leave that out. If the worry is that a CME may occur from the set being added to while looping through it, a simple `boolean` state keeper is better here.

```java
boolean processing = false;

public static void schedulePlayerInventoryCheck(EntityPlayer player) {
    if (processing) {
        return;
    }
}

// ...

public void onServerTick(ServerTickEvent event) {
    if (event.phase == Phase.START) {
        processing = true;
        // Logic
        processing = false;
    }
}
```